kernel/kernel_linux.go

```diff
@@ -0,0 +1,28 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package kernel
+
+import "golang.org/x/sys/unix"
+
+// swapDirsAtomically swaps atomically two directories by using
+// renameat2 syscall. The directories must be absolute.
+func swapDirsAtomically(oldpath string, newpath string) (err error) {
+	return unix.Renameat2(-1, oldpath, -1, newpath, unix.RENAME_EXCHANGE)
```

Comment:
wonder if this needs a sync on  `filepath.Dir(newpath)` directory maybe.

Reply:
It does probably, but as commented I prefer to do a single sync after the two swaps. It is a bit of an issue that here "atomic" really refers to the running system so any access to files under the directory after the swap would find the content of the new directory, not to the filesystem, so in the end I have changed the name to `SwapDirs` to avoid that word.